h/traversal/annotation.py

```diff
@@ -93,4 +94,4 @@ def _group_principals(group, permission):
         if group is None:
             return []
 
-        return principals_allowed_by_permission(group, permission)
+        return principals_allowed_by_permission(GroupContext(group), permission)
```

Comment:
The previous thousand lines of diff in two other PRs or so are to make this change possible...

It's almost entirely down to the tests being impossible to maintain.